[🐛 Bug]: Timeout error with reloadSession API

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

latest

### Node.js Version

20

### Mode

Standalone Mode

### Which capabilities are you using?

```typescript
{
    browserName: 'chrome',
    acceptInsecureCerts: true,
    'goog:chromeOptions': {
      args: [
        '--disable-backgrounding-occluded-windows',
        '--window-size=1980,1080',
        '--lang=fr_FR.UTF-8',
        '--disable-search-engine-choice-screen',
      ],
    },
  },

Grid with Selenium-server-4.25.0
```


### What happened?

If reload session is used, the test breaks with a timeout error.

### What is your expected behavior?

_No response_

### How to reproduce the bug.

```
  await browser.reloadSession();
  await browser.url(
    'https://webdriver.io/blog/2024/08/15/webdriverio-v9-release'
  );
  await $('[href="/blog"]').waitForDisplayed();
```

### Relevant log output

```typescript
[0-0] 2024-10-22T13:55:30.142Z INFO webdriver: COMMAND deleteSession(<object>)
[0-0] 2024-10-22T13:55:30.143Z INFO webdriver: [DELETE] http://127.0.0.1:4444/wd/hub/session/bff0845355593ed7b100644eaccc694f
[0-0] 2024-10-22T13:55:30.143Z INFO webdriver: DATA { deleteSessionOpts: { shutdownDriver: false } }
[0-0] 2024-10-22T13:55:30.809Z INFO webdriver: RESULT null
[0-0] 2024-10-22T13:55:30.812Z INFO webdriver: [POST] http://127.0.0.1:4444/wd/hub/session
[0-0] 2024-10-22T13:55:30.812Z INFO webdriver: DATA {
[0-0]   capabilities: {
[0-0]     alwaysMatch: {
[0-0]       browserName: 'chrome',
[0-0]       acceptInsecureCerts: true,
[0-0]       'goog:chromeOptions': [Object],
[0-0]       webSocketUrl: true
[0-0]     },
[0-0]     firstMatch: [ {} ]
[0-0]   }
[0-0] }
[0-0] 2024-10-22T13:55:33.211Z INFO webdriver: BIDI COMMAND browsingContext.navigate {"context":"CE6C5F1343C782BBC3DED3A225FE0C50","url":"https://webdriver.io/blog/2024/08/15/webdriverio-v9-release","wait":"complete"}
[0-0] 2024-10-22T13:56:33.257Z INFO webdriver: COMMAND getUrl()
[0-0] 2024-10-22T13:56:33.258Z INFO webdriver: [GET] http://127.0.0.1:4444/wd/hub/session/cdbf45f49986aea09d6a6f435e50e69d/url
[0-0] 2024-10-22T13:56:33.278Z INFO webdriver: RESULT about:blank
[0-0] Error: Command browsingContext.navigate with id 9 (with the following parameter: {"context":"CE6C5F1343C782BBC3DED3A225FE0C50","url":"https://webdriver.io/blog/2024/08/15/webdriverio-v9-release","wait":"complete"}) timed out
    at Timeout._onTimeout (C:\2-gitlab\end-to-end-tests-wdio\node_modules\webdriver\build\index.js:246:16)
    at listOnTimeout (node:internal/timers:581:17)
    at process.processTimers (node:internal/timers:519:7)
```
Selenium grid errors
```
 Session request received by the Distributor:
 [Capabilities {acceptInsecureCerts: true, browserName: chrome, goog:chromeOptions: {args: [--disable-backgrounding-occ..., --window-size=1980,1080, --lang=fr_FR.UTF-8, --disable-search-engine-cho...], prefs: {}}, webSocketUrl: true}]
10:11:50.268 INFO [LocalNode.newSession] - Session created by the Node. Id: e65e05edb82b40d7ed9cc92839a42efd, Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 130.0.6723.58, chrome: {chromedriverVersion: 130.0.6723.69 (3ec172b971b9..., userDataDir: C:\Users\FABIEN~1\AppData\L...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:62413}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: Windows 10, proxy: Proxy(), se:cdp: ws://172.23.176.1:4444/sess..., se:cdpVersion: 130.0.6723.58, se:gridWebSocketUrl: ws://localhost:14852/sessio..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webSocketUrl: ws://172.23.176.1:4444/sess..., webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
10:11:50.277 INFO [LocalDistributor.newSession] - Session created by the Distributor. Id: e65e05edb82b40d7ed9cc92839a42efd
 Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 130.0.6723.58, chrome: {chromedriverVersion: 130.0.6723.69 (3ec172b971b9..., userDataDir: C:\Users\FABIEN~1\AppData\L...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:62413}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: Windows 10, proxy: Proxy(), se:cdp: ws://172.23.176.1:4444/sess..., se:cdpVersion: 130.0.6723.58, se:gridWebSocketUrl: ws://localhost:14852/sessio..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webSocketUrl: ws://172.23.176.1:4444/sess..., webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
10:11:50.334 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:14852/session/e65e05edb82b40d7ed9cc92839a42efd
10:12:00.244 INFO [LocalSessionMap.remove] - Deleted session from local Session Map, Id: e65e05edb82b40d7ed9cc92839a42efd
10:12:00.247 INFO [GridModel.release] - Releasing slot for session id e65e05edb82b40d7ed9cc92839a42efd
10:12:00.253 INFO [SessionSlot.stop] - Stopping session e65e05edb82b40d7ed9cc92839a42efd
10:12:00.282 INFO [LocalDistributor.newSession] - Session request received by the Distributor:
 [Capabilities {acceptInsecureCerts: true, browserName: chrome, goog:chromeOptions: {args: [--disable-backgrounding-occ..., --window-size=1980,1080, --lang=fr_FR.UTF-8, --disable-search-engine-cho...], prefs: {}}, webSocketUrl: true}]
10:12:02.484 INFO [LocalNode.newSession] - Session created by the Node. Id: 86f60d538758d2e1472e2934bd45bb25, Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 130.0.6723.58, chrome: {chromedriverVersion: 130.0.6723.69 (3ec172b971b9..., userDataDir: C:\Users\FABIEN~1\AppData\L...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:62469}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: Windows 10, proxy: Proxy(), se:cdp: ws://172.23.176.1:4444/sess..., se:cdpVersion: 130.0.6723.58, se:gridWebSocketUrl: ws://localhost:32182/sessio..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webSocketUrl: ws://172.23.176.1:4444/sess..., webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
10:12:02.490 INFO [LocalDistributor.newSession] - Session created by the Distributor. Id: 86f60d538758d2e1472e2934bd45bb25
 Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 130.0.6723.58, chrome: {chromedriverVersion: 130.0.6723.69 (3ec172b971b9..., userDataDir: C:\Users\FABIEN~1\AppData\L...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:62469}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: Windows 10, proxy: Proxy(), se:cdp: ws://172.23.176.1:4444/sess..., se:cdpVersion: 130.0.6723.58, se:gridWebSocketUrl: ws://localhost:32182/sessio..., setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webSocketUrl: ws://172.23.176.1:4444/sess..., webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
10:12:02.500 WARN [DefaultChannelPipeline.onUnhandledInboundException] - An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
org.openqa.selenium.WebDriverException: java.io.IOException: Output closed
Build info: version: '4.25.0', revision: '030fcf7918'
System info: os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '15.0.1'
Driver info: driver.version: unknown
        at org.openqa.selenium.remote.http.jdk.JdkHttpClient$4.send(JdkHttpClient.java:320)
        at org.openqa.selenium.grid.node.ProxyNodeWebsockets.lambda$createWsEndPoint$6(ProxyNodeWebsockets.java:237)
        at org.openqa.selenium.netty.server.WebSocketMessageHandler.lambda$channelRead0$0(WebSocketMessageHandler.java:47)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: Output closed
        at java.net.http/jdk.internal.net.http.websocket.MessageEncoder.encodeText(MessageEncoder.java:136)
        at java.net.http/jdk.internal.net.http.websocket.TransportImpl$SendTask$1.onText(TransportImpl.java:366)
        at java.net.http/jdk.internal.net.http.websocket.TransportImpl$SendTask$1.onText(TransportImpl.java:357)
        at java.net.http/jdk.internal.net.http.websocket.MessageQueue.peek(MessageQueue.java:223)
        at java.net.http/jdk.internal.net.http.websocket.TransportImpl$SendTask.run(TransportImpl.java:545)
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:147)
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:198)
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:271)
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:224)
        at java.net.http/jdk.internal.net.http.websocket.TransportImpl.sendText(TransportImpl.java:149)
        at java.net.http/jdk.internal.net.http.websocket.WebSocketImpl.sendText(WebSocketImpl.java:184)
        at org.openqa.selenium.remote.http.jdk.JdkHttpClient$4.lambda$send$1(JdkHttpClient.java:271)
        at org.openqa.selenium.remote.http.jdk.JdkHttpClient$4.send(JdkHttpClient.java:310)
        ... 10 more
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

--- a/e2e/reloadSession.test.ts
+++ b/e2e/reloadSession.test.ts
@@ -1,0 +1,25 @@
+import { test, expect } from 'vitest'
+/**
+ * in order to run this file make sure you have `webdriverio`
+ * installed using NPM before running it:
+ *
+ *   $ npm install webdriverio
+ *
+ */
+import { remote } from 'webdriverio'
+
+test('can reconnect to WebDriver Bidi session', async () => {
+    const browser = await remote({
+        capabilities: {
+            browserName: 'chrome',
+            'goog:chromeOptions': {
+                args: ['headless', 'disable-gpu']
+            }
+        }
+    })
+
+    expect(typeof await browser.browsingContextGetTree({})).toBe('object')
+    await browser.reloadSession()
+    expect(typeof await browser.browsingContextGetTree({})).toBe('object')
+    await browser.deleteSession()
+})

--- a/e2e/wdio/headless/puppeteer.e2e.ts
+++ b/e2e/wdio/headless/puppeteer.e2e.ts
@@ -1,7 +1,7 @@
-import { browser } from '../../../packages/wdio-globals/build/index.js'
+import { browser, expect } from '@wdio/globals'
 
 describe('WebdriverIO', () => {
-    it.skip('should provide access to Puppeteer', async () => {
+    it('should provide access to Puppeteer', async () => {
         const puppeteerInstance = await browser.getPuppeteer()
         expect(puppeteerInstance).toBeDefined()
     })

--- a/e2e/wdio/headless/reloadSession.e2e.ts
+++ b/e2e/wdio/headless/reloadSession.e2e.ts
@@ -1,13 +1,8 @@
 import { browser, expect } from '@wdio/globals'
 
 describe('reloadSession', () => {
-    it('can reload a session', async () => {
-        const sessionId = browser.sessionId
-        await browser.reloadSession()
-        expect(browser.sessionId).not.toBe(sessionId)
-    })
-
     it('can reload a session with new capabilities', async () => {
+        const sessionId = browser.sessionId
         expect(browser.capabilities.browserName).toBe('chrome')
         await browser.reloadSession({
             browserName: 'edge',
@@ -16,5 +11,6 @@ describe('reloadSession', () => {
             }
         })
         expect(browser.capabilities.browserName).toContain('MicrosoftEdge')
+        expect(browser.sessionId).not.toBe(sessionId)
     })
 })

--- a/e2e/wdio/headless/source-maps.e2e.ts
+++ b/e2e/wdio/headless/source-maps.e2e.ts
@@ -1,7 +1,8 @@
+import { expect } from '@wdio/globals'
 import { Task } from '@serenity-js/core'
 
 // eslint-disable-next-line no-undef
-describe.skip('Source maps support for ESM projects', () => {
+describe('Source maps support for ESM projects', () => {
 
     // eslint-disable-next-line no-undef
     it('should detect line number correctly', async () => {
@@ -13,15 +14,15 @@ describe.skip('Source maps support for ESM projects', () => {
         const locations = frames.map(frame => frame.match(/.*\/(.*?\d+:\d+)/)[1]).slice(0, 4)
 
         expect(locations).toEqual([
-            'source-maps.e2e.ts:43:19',   // the line where we instantiate the Error object
-            'source-maps.e2e.ts:39:12',   // fn2 calls readStack
-            'source-maps.e2e.ts:35:12',   // fn1 calls fn2
-            'source-maps.e2e.ts:8:38',    // fn1 called in the spec
+            'source-maps.e2e.ts:44:19',   // the line where we instantiate the Error object
+            'source-maps.e2e.ts:40:12',   // fn2 calls readStack
+            'source-maps.e2e.ts:36:12',   // fn1 calls fn2
+            'source-maps.e2e.ts:9:38',    // fn1 called in the spec
         ])
     })
 
     // eslint-disable-next-line no-undef
-    it('should detect instantiation location of a Serenity/JS interaction', async () => {
+    it.skip('should detect instantiation location of a Serenity/JS interaction', async () => {
 
         const location = serenityTask()     // <-- Instantiation location line should point at line 26
             .instantiationLocation()

--- a/e2e/wdio/headless/source-maps.e2e.ts
+++ b/e2e/wdio/headless/source-maps.e2e.ts
@@ -2,7 +2,7 @@ import { expect } from '@wdio/globals'
 import { Task } from '@serenity-js/core'
 
 // eslint-disable-next-line no-undef
-describe('Source maps support for ESM projects', () => {
+describe.skip('Source maps support for ESM projects', () => {
 
     // eslint-disable-next-line no-undef
     it('should detect line number correctly', async () => {
@@ -22,7 +22,7 @@ describe('Source maps support for ESM projects', () => {
     })
 
     // eslint-disable-next-line no-undef
-    it.skip('should detect instantiation location of a Serenity/JS interaction', async () => {
+    it('should detect instantiation location of a Serenity/JS interaction', async () => {
 
         const location = serenityTask()     // <-- Instantiation location line should point at line 26
             .instantiationLocation()

--- a/packages/webdriver/src/bidi/core.ts
+++ b/packages/webdriver/src/bidi/core.ts
@@ -28,6 +28,15 @@ export class BidiCore {
         this.#ws.on('message', this.#handleResponse.bind(this))
     }
 
+    /**
+     * We initiate the Bidi instance before a WebdriverIO instance is created.
+     * In order to emit Bidi events we have to attach the WebdriverIO instance
+     * to the Bidi instance afterwards.
+     */
+    public attachClient (client: Client) {
+        this.client = client
+    }
+
     public async connect () {
         /**
          * don't connect and stale unit tests when the websocket url is set to a dummy value
@@ -43,6 +52,27 @@ export class BidiCore {
             resolve()
         }))
         return this.#waitForConnected
+    }
+
+    public close () {
+        if (!this.#isConnected) {
+            return
+        }
+
+        log.info(`Close Bidi connection to ${this.#webSocketUrl}`)
+        this.#isConnected = false
+        this.#ws.off('message', this.#handleResponse.bind(this))
+        this.#ws.close()
+        this.#ws.terminate()
+    }
+
+    public reconnect (webSocketUrl: string, opts?: ClientOptions) {
+        log.info(`Reconnect to new Bidi session at ${webSocketUrl}`)
+        this.close()
+        this.#webSocketUrl = webSocketUrl
+        this.#ws = new Socket(this.#webSocketUrl, opts) as WebSocket
+        this.#ws.on('message', this.#handleResponse.bind(this))
+        return this.connect()
     }
 
     /**
@@ -76,7 +106,7 @@ export class BidiCore {
                 return
             }
 
-            log.debug('BIDI RESULT', data.toString())
+            log.info('BIDI RESULT', data.toString())
             this.client?.emit('bidiResult', payload)
             const resolve = this.#pendingCommands.get(payload.id)
             if (!resolve) {

--- a/packages/webdriver/src/command.ts
+++ b/packages/webdriver/src/command.ts
@@ -155,9 +155,13 @@ export default function (
                  */
                 if (shutdownDriver && 'wdio:driverPID' in this.capabilities && this.capabilities['wdio:driverPID']) {
                     log.info(`Kill driver process with PID ${this.capabilities['wdio:driverPID']}`)
-                    const killedSuccessfully = process.kill(this.capabilities['wdio:driverPID'], 'SIGKILL')
-                    if (!killedSuccessfully) {
-                        log.warn('Failed to kill driver process, manually clean-up might be required')
+                    try {
+                        const killedSuccessfully = process.kill(this.capabilities['wdio:driverPID'], 'SIGKILL')
+                        if (!killedSuccessfully) {
+                            log.warn('Failed to kill driver process, manually clean-up might be required')
+                        }
+                    } catch (err) {
+                        log.warn('Failed to kill driver process', err)
                     }
 
                     setTimeout(() => {

--- a/packages/webdriver/src/command.ts
+++ b/packages/webdriver/src/command.ts
@@ -3,6 +3,7 @@ import { commandCallStructure, isValidParameter, getArgumentType } from '@wdio/u
 import { WebDriverBidiProtocol, type CommandEndpoint } from '@wdio/protocols'
 
 import Request from './request/request.js'
+import type { BidiHandler } from './bidi/handler.js'
 import type { WebDriverResponse } from './request/types.js'
 import type { BaseClient, BidiCommands, BidiResponses } from './types.js'
 
@@ -142,6 +143,12 @@ export default function (
             this.emit('result', { command, method, endpoint, body, result })
 
             if (command === 'deleteSession') {
+                /**
+                 * close WebDriver Bidi socket
+                 */
+                const browser = this as { _bidiHandler?: BidiHandler }
+                browser._bidiHandler?.close()
+
                 const shutdownDriver = body.deleteSessionOpts?.shutdownDriver !== false
                 /**
                  * kill driver process if there is one

--- a/packages/webdriver/src/index.ts
+++ b/packages/webdriver/src/index.ts
@@ -8,6 +8,7 @@ import type { Capabilities, Options } from '@wdio/types'
 
 import command from './command.js'
 import { DEFAULTS } from './constants.js'
+import type { BidiHandler } from './bidi/handler.js'
 import { startWebDriverSession, getPrototype, getEnvironmentVars, setupDirectConnect, initiateBidi, parseBidiMessage } from './utils.js'
 import type { Client, AttachOptions, SessionFlags } from './types.js'
 
@@ -150,7 +151,7 @@ export default class WebDriver {
      * @param   {object} instance  the object we get from a new browser session.
      * @returns {string}           the new session id of the browser
      */
-    static async reloadSession(instance: Client, newCapabilities?: WebdriverIO.Capabilities) {
+    static async reloadSession(instance: Client & { _bidiHandler?: BidiHandler }, newCapabilities?: WebdriverIO.Capabilities) {
         const capabilities = newCapabilities ? newCapabilities : Object.assign({}, instance.requestedCapabilities) as WebdriverIO.Capabilities
         let params: Capabilities.RemoteConfig = { ...instance.options, capabilities }
 
@@ -190,9 +191,21 @@ export default class WebDriver {
         for (const prop in instance.requestedCapabilities) {
             delete instance.requestedCapabilities[prop as keyof typeof instance.requestedCapabilities]
         }
+
+        const driverPid = instance.capabilities['wdio:driverPID']
         instance.sessionId = sessionId
         instance.capabilities = newSessionCapabilities
+        instance.capabilities['wdio:driverPID'] = driverPid
         Object.assign(instance.requestedCapabilities, capabilities)
+
+        /**
+         * reconnect to new Bidi session
+         */
+        if (isBidi(instance.requestedCapabilities || {}, instance.capabilities || {})) {
+            const bidiReqOpts = instance.options.strictSSL ? {} : { rejectUnauthorized: false }
+            await instance._bidiHandler?.reconnect(newSessionCapabilities.webSocketUrl as unknown as string, bidiReqOpts)
+        }
+
         return sessionId
     }
 

--- a/packages/webdriver/src/utils.ts
+++ b/packages/webdriver/src/utils.ts
@@ -378,7 +378,7 @@ export function initiateBidi (socketUrl: string, strictSSL: boolean = true): Pro
                     /**
                      * attach the client to the handler to emit events
                      */
-                    handler.client = this
+                    handler.attachClient(this)
 
                     this.emit(cur.command, args)
                     return bidiFn?.apply(handler, args)

